pointgrey_camera_driver/src/nodelet.cpp

```diff
@@ -312,73 +273,187 @@ class PointGreyCameraNodelet: public nodelet::Nodelet
-        // Get the image from the camera library
-        NODELET_DEBUG("Starting a new grab from camera.");
-        pg_.grabImage(wfov_image->image, frame_id_);
-
-        // Set other values
-        wfov_image->header.frame_id = frame_id_;
-
-        wfov_image->gain = gain_;
-        wfov_image->white_balance_blue = wb_blue_;
-        wfov_image->white_balance_red = wb_red_;
-
-        wfov_image->temperature = pg_.getCameraTemperature();
-
-        ros::Time time = ros::Time::now();
-        wfov_image->header.stamp = time;
-        wfov_image->image.header.stamp = time;
-
-        // Set the CameraInfo message
-        ci_.reset(new sensor_msgs::CameraInfo(cinfo_->getCameraInfo()));
-        ci_->header.stamp = wfov_image->image.header.stamp;
-        ci_->header.frame_id = wfov_image->header.frame_id;
-        // The height, width, distortion model, and parameters are all filled in by camera info manager.
-        ci_->binning_x = binning_x_;
-        ci_->binning_y = binning_y_;
-        ci_->roi.x_offset = roi_x_offset_;
-        ci_->roi.y_offset = roi_y_offset_;
-        ci_->roi.height = roi_height_;
-        ci_->roi.width = roi_width_;
-        ci_->roi.do_rectify = do_rectify_;
-
-        wfov_image->info = *ci_;
-
-        // Publish the full message
-        pub_->publish(wfov_image);
-
-        // Publish the message using standard image transport
-        if(it_pub_.getNumSubscribers() > 0)
-        {
-          sensor_msgs::ImagePtr image(new sensor_msgs::Image(wfov_image->image));
-          it_pub_.publish(image, ci_);
-        }
+      bool state_changed = state != previous_state;
 
+      previous_state = state;
 
-      }
-      catch(CameraTimeoutException& e)
+      switch(state)
       {
-        NODELET_WARN("%s", e.what());
-      }
-      catch(std::runtime_error& e)
-      {
-        NODELET_ERROR("%s", e.what());
-        ///< @todo Look into readding this
-        /*try{
-          // Something terrible has happened, so let's just disconnect and reconnect to see if we can recover.
-          pg_.disconnect();
-          ros::Duration(1.0).sleep(); // sleep for one second each time
-          pg_.connect();
-          pg_.start();
-        } catch(std::runtime_error& e2){
-          NODELET_ERROR("%s", e2.what());
-        }*/
+        case ERROR:
+          // Generally there's no need to stop before disconnecting after an
+          // error. Indeed, stop will usually fail.
+#if STOP_ON_ERROR
+          // Try stopping the camera
+          try
+          {
+            sub_.shutdown();
+
+            NODELET_DEBUG("Stopping camera.");
+            pg_.stop();
+            NODELET_INFO("Stopped camera.");
+
+            state = STOPPED;
+          }
+          catch(std::runtime_error& e)
+          {
+            NODELET_ERROR_COND(state_changed,
+                "Failed to stop error: %s", e.what());
+            ros::Duration(1.0).sleep(); // sleep for one second each time
```

Comment:
The only two `catch` blocks w/o sleep are the original ones, which now are like this:

``` c++
          catch(CameraTimeoutException& e)
          {
            NODELET_WARN("%s", e.what());
          }
          catch(std::runtime_error& e)
          {
            NODELET_ERROR("%s", e.what());

            state = ERROR;
          }
```

For the first one, since it's a timeout exception, I don't think we need to sleep. It won't spam a full speed because that exception can only happen on timeout.

For the second one, we enter `ERROR` state, so there's no problem on iterating immediately and do the error recovery w/o waiting. Maybe waiting we save on error message sometimes, when we try to disconnect from the camera, but only sometimes.